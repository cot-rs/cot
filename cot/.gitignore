# Generated files
static/admin/admin.css
templates/css/error.css
