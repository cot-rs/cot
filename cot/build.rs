use std::path::PathBuf;

#[cfg(all(
    feature = "db",
    not(any(feature = "sqlite", feature = "postgres", feature = "mysql"))
))]
compile_error!("feature \"db\" requires one of: \"sqlite\", \"postgres\", \"mysql\" to be enabled");

fn main() {
    build_css();
}

fn build_css() {
    const SCSS_FILES: [(&str, &str); 2] = [
        ("admin/admin.scss", "static/admin/admin.css"),
        ("error.scss", "templates/css/error.css"),
    ];

    let options = scss_options();

    for (scss_file, css_file) in SCSS_FILES {
        let scss_path = format!("scss/{scss_file}");

        println!("cargo::rerun-if-changed={scss_path}");

        let css = grass::from_path(scss_path, &options).expect("failed to compile SCSS");

        let out_dir = std::env::var("OUT_DIR").expect("OUT_DIR should be set");
        let css_path = PathBuf::from(out_dir).join(css_file);
        let css_dir = css_path
            .parent()
            .expect("failed to get CSS parent directory");
        std::fs::create_dir_all(css_dir).expect("failed to create CSS directory");
        std::fs::write(css_path, css).expect("failed to write CSS");
    }
}

fn scss_options() -> grass::Options<'static> {
    grass::Options::default().style(grass::OutputStyle::Compressed)
}
