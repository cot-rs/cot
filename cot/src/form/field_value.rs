use bytes::Bytes;
use thiserror::Error;

/// A value from a form field.
///
/// This type represents a value from a form field, which can be either a text
/// value or a multipart field (like a file upload). It provides methods to
/// access the field's metadata (name, filename, content type) and to convert
/// the value into different formats.
#[derive(Debug)]
pub struct FormFieldValue<'a> {
    inner: FormFieldValueImpl<'a>,
}

impl<'a> FormFieldValue<'a> {
    /// Creates a new text field value.
    ///
    /// # Examples
    ///
    /// ```
    /// use cot::form::FormFieldValue;
    ///
    /// # #[tokio::main]
    /// # async fn main() -> cot::Result<()> {
    /// let value = FormFieldValue::new_text("hello");
    /// assert_eq!(value.is_multipart(), false);
    /// assert_eq!(value.filename(), None);
    /// assert_eq!(value.content_type(), None);
    /// assert_eq!(value.into_text().await?, "hello");
    /// # Ok(())
    /// # }
    /// ```
    #[must_use]
    pub fn new_text<T: Into<String>>(text: T) -> Self {
        Self {
            inner: FormFieldValueImpl::Text(text.into()),
        }
    }

    /// Creates a new multipart field value.
    #[must_use]
    pub(crate) fn new_multipart(field: multer::Field<'a>) -> Self {
        Self {
            inner: FormFieldValueImpl::Multipart(Box::new(MultipartField { inner: field })),
        }
    }

    /// Returns the filename of the field, if it has one.
    ///
    /// Only multipart fields can have filenames. Text fields always return
    /// `None`.
    #[must_use]
    pub fn filename(&self) -> Option<&str> {
        match &self.inner {
            FormFieldValueImpl::Text(_) => None,
            FormFieldValueImpl::Multipart(multipart) => multipart.inner.file_name(),
        }
    }

    /// Returns the content type of the field, if it has one.
    ///
    /// Only multipart fields can have content types. Text fields always return
    /// `None`.
    #[must_use]
    pub fn content_type(&self) -> Option<&str> {
        match &self.inner {
            FormFieldValueImpl::Text(_) => None,
            FormFieldValueImpl::Multipart(multipart) => {
                multipart.inner.content_type().map(AsRef::as_ref)
            }
        }
    }

    /// Converts the field value into bytes.
    ///
    /// For text fields, this converts the text into bytes. For multipart
    /// fields, this reads the field's content as bytes.
    ///
    /// # Examples
    ///
    /// ```
    /// use bytes::Bytes;
    /// use cot::form::FormFieldValue;
    ///
    /// # #[tokio::main]
    /// # async fn main() -> cot::Result<()> {
    /// let text_value = FormFieldValue::new_text("hello");
    /// assert_eq!(text_value.into_bytes().await?, Bytes::from("hello"));
    /// # Ok(())
    /// # }
    /// ```
    pub async fn into_bytes(self) -> Result<Bytes, FormFieldValueError> {
        match self.inner {
            FormFieldValueImpl::Text(text) => Ok(Bytes::from(text)),
            FormFieldValueImpl::Multipart(multipart) => multipart
                .inner
                .bytes()
                .await
                .map_err(FormFieldValueError::from_multer),
        }
    }

    /// Converts the field value into text.
    ///
    /// For text fields, this returns the text directly. For multipart fields,
    /// this reads the field's content as text.
    ///
    /// # Examples
    ///
    /// ```
    /// use cot::form::FormFieldValue;
    ///
    /// # #[tokio::main]
    /// # async fn main() -> cot::Result<()> {
    /// let text_value = FormFieldValue::new_text("hello");
    /// assert_eq!(text_value.into_text().await?, "hello");
    /// # Ok(())
    /// # }
    /// ```
    pub async fn into_text(self) -> Result<String, FormFieldValueError> {
        match self.inner {
            FormFieldValueImpl::Text(text) => Ok(text),
            FormFieldValueImpl::Multipart(multipart) => multipart
                .inner
                .text()
                .await
                .map_err(FormFieldValueError::from_multer),
        }
    }

    /// Returns whether this is a multipart field.
    ///
    /// # Examples
    ///
    /// ```
    /// use cot::form::FormFieldValue;
    ///
    /// let text_value = FormFieldValue::new_text("hello");
    /// assert!(!text_value.is_multipart());
    /// ```
    #[must_use]
    pub fn is_multipart(&self) -> bool {
        matches!(self.inner, FormFieldValueImpl::Multipart(_))
    }
}

#[derive(Debug)]
enum FormFieldValueImpl<'a> {
    Text(String),
    Multipart(Box<MultipartField<'a>>),
}

#[derive(Debug)]
struct MultipartField<'a> {
    inner: multer::Field<'a>,
}

/// An error that can occur when processing a form field value.
///
/// This type represents errors that can occur when processing form field
/// values, such as errors from the multipart parser or validation errors.
#[derive(Debug, PartialEq, Eq, Error)]
#[error(transparent)]
pub struct FormFieldValueError {
    inner: FormFieldValueErrorImpl,
}

#[derive(Debug, PartialEq, Eq, Error)]
enum FormFieldValueErrorImpl {
    #[error(transparent)]
    Multer(multer::Error),
    #[error("Multipart field does not have a name")]
    NoName,
    #[error("File field requires the form to be sent as multipart/form-data")]
    MultipartRequired,
}

impl FormFieldValueError {
    pub(crate) fn from_multer(multer: multer::Error) -> Self {
        Self {
            inner: FormFieldValueErrorImpl::Multer(multer),
        }
    }

    pub(crate) fn no_name() -> Self {
        Self {
            inner: FormFieldValueErrorImpl::NoName,
        }
    }

    pub(crate) fn multipart_required() -> Self {
        Self {
            inner: FormFieldValueErrorImpl::MultipartRequired,
        }
    }
}

#[cfg(test)]
mod tests {
    use bytes::Bytes;
    use futures_util::stream::once;
    use multer::Multipart;

    use super::*;

    #[cot::test]
    async fn text_field_value() {
        let value = FormFieldValue::new_text("hello");

        assert!(!value.is_multipart());
        assert_eq!(value.filename(), None);
        assert_eq!(value.content_type(), None);
        assert_eq!(value.into_text().await.unwrap(), "hello");

        let value = FormFieldValue::new_text("hello");

        assert_eq!(value.into_bytes().await.unwrap(), Bytes::from("hello"));
    }

    #[cot::test]
    async fn multipart_field_value() {
        let boundary = "boundary";
        let body = format!(
            "--{boundary}\r\n\
            Content-Disposition: form-data; name=\"file\"; filename=\"test.txt\"\r\n\
            Content-Type: text/plain\r\n\
            \r\n\
            file content\r\n\
            --{boundary}--\r\n"
        );

        let value = test_multipart_value(boundary, body.clone()).await;
        assert!(value.is_multipart());
        assert_eq!(value.filename(), Some("test.txt"));
        assert_eq!(value.content_type(), Some("text/plain"));
        assert_eq!(value.into_text().await.unwrap(), "file content");

        let value = test_multipart_value(boundary, body).await;
        assert_eq!(value.into_bytes().await.unwrap(), "file content");
    }

    async fn test_multipart_value(boundary: &str, body: String) -> FormFieldValue<'static> {
        let stream = once(async move { Ok::<_, std::io::Error>(Bytes::from(body)) });
        let mut multipart = Multipart::new(stream, boundary);

        let field = multipart.next_field().await.unwrap().unwrap();
        let value = FormFieldValue::new_multipart(field);

        value
    }
}
