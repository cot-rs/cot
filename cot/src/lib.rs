//! Cot is an easy to use, modern, and fast web framework for Rust. It has
//! been designed to be familiar if you've ever used
//! [Django](https://www.djangoproject.com/), and easy to learn if you haven't.
//! It's a batteries-included framework built on top of
//! [axum](https://github.com/tokio-rs/axum).
//!
//! ## Features
//!
//! * **Easy to use API** — in many ways modeled after Django, Cot's API is
//!   designed to be easy to use and intuitive. Sensible defaults make it for
//!   easy rapid development, while the API is still empowering you when needed.
//!   The documentation is a first-class citizen in Cot, making it easy to find
//!   what you're looking for.
//! * **ORM integration** — Cot comes with its own ORM, allowing you to interact
//!   with your database in a way that feels Rusty and intuitive. Rust types are
//!   the source of truth, and the ORM takes care of translating them to and
//!   from the database, as well as creating the migrations automatically.
//! * **Type safe** — wherever possible, Cot uses Rust's type system to prevent
//!   common mistakes and bugs. Not only views are taking advantage of the
//!   Rust's type system, but also the ORM, the admin panel, and even the
//!   templates. All that to catch errors as early as possible.
//! * **Admin panel** — Cot comes with an admin panel out of the box, allowing
//!   you to manage your app's data with ease. Adding new models to the admin
//!   panel is stupidly simple, making it a great tool not only for rapid
//!   development and debugging, but with its customization options, also for
//!   production use.
//! * **Secure by default** — security should be opt-out, not opt-in. Cot takes
//!   care of making your web apps secure by default, defending it against
//!   common modern web vulnerabilities. You can focus on building your app, not
//!   securing it.
//!
//! ## Guide
//!
//! This is an API reference for Cot, which might not be the best place to
//! start learning Cot. For a more gentle introduction, see the
//! [Cot guide](https://cot.rs/guide/latest/).
//!
//! ## Examples
//!
//! To see examples of how to use Cot, see the
//! [examples in the repository](https://github.com/cot-rs/cot/tree/master/examples).

#![warn(missing_docs, rustdoc::missing_crate_level_docs)]
#![cfg_attr(
    docsrs,
    feature(doc_cfg, doc_auto_cfg, rustdoc_missing_doc_code_examples)
)]
#![cfg_attr(docsrs, warn(rustdoc::missing_doc_code_examples))]
#![doc(
    html_logo_url = "https://raw.githubusercontent.com/cot-rs/media/6585c518/logo/logo.svg",
    html_favicon_url = "https://raw.githubusercontent.com/cot-rs/media/6585c518/logo/favicon.svg"
)]

extern crate self as cot;

#[cfg(feature = "db")]
pub mod db;
mod error;
pub mod form;
mod headers;
// Not public API. Referenced by macro-generated code.
#[doc(hidden)]
#[path = "private.rs"]
pub mod __private;
pub mod admin;
pub mod auth;
mod body;
pub mod cli;
pub mod config;
mod error_page;
mod handler;
pub mod html;
pub mod middleware;
#[cfg(feature = "openapi")]
pub mod openapi;
pub mod project;
pub mod request;
pub mod response;
pub mod router;
pub mod session;
pub mod static_files;
pub mod test;
pub(crate) mod utils;

#[cfg(feature = "openapi")]
pub use aide;
pub use body::Body;
pub use cot_macros::{main, test};
pub use error::Error;
#[cfg(feature = "openapi")]
pub use schemars;
pub use {bytes, http};

pub use crate::handler::{BoxedHandler, RequestHandler};
pub use crate::project::{
    App, AppBuilder, Bootstrapper, Project, ProjectContext, run, run_at, run_cli,
};

/// A type alias for a result that can return a [`cot::Error`].
pub type Result<T> = std::result::Result<T, Error>;

/// A type alias for an HTTP status code.
pub type StatusCode = http::StatusCode;

/// A type alias for an HTTP method.
pub type Method = http::Method;
