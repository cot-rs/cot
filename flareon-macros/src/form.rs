use std::collections::HashMap;

use darling::{FromDeriveInput, FromField};
use proc_macro2::TokenStream;
use quote::{format_ident, quote, ToTokens, TokenStreamExt};

use crate::flareon_ident;

pub(super) fn impl_form_for_struct(ast: &syn::DeriveInput) -> TokenStream {
    let opts = match FormOpts::from_derive_input(ast) {
        Ok(val) => val,
        Err(err) => {
            return err.write_errors();
        }
    };

    let mut builder = opts.as_form_derive_builder();
    for field in opts.fields() {
        builder.push_field(field);
    }

    quote!(#builder)
}

#[derive(Debug, FromDeriveInput)]
#[darling(forward_attrs(allow, doc, cfg), supports(struct_named))]
struct FormOpts {
    ident: syn::Ident,
    data: darling::ast::Data<darling::util::Ignored, Field>,
}

impl FormOpts {
    fn fields(&self) -> Vec<&Field> {
        self.data
            .as_ref()
            .take_struct()
            .expect("Only structs are supported")
            .fields
    }

    fn field_count(&self) -> usize {
        self.fields().len()
    }

    fn as_form_derive_builder(&self) -> FormDeriveBuilder {
        FormDeriveBuilder {
            name: self.ident.clone(),
            context_struct_name: format_ident!("{}Context", self.ident),
            context_struct_errors_name: format_ident!("{}ContextErrors", self.ident),
            fields_as_struct_fields: Vec::with_capacity(self.field_count()),
            fields_as_struct_fields_new: Vec::with_capacity(self.field_count()),
            fields_as_context_from_request: Vec::with_capacity(self.field_count()),
            fields_as_from_context_vars: Vec::with_capacity(self.field_count()),
            fields_as_from_context: Vec::with_capacity(self.field_count()),
            fields_as_errors: Vec::with_capacity(self.field_count()),
            fields_as_errors_for: Vec::with_capacity(self.field_count()),
            fields_as_errors_for_mut: Vec::with_capacity(self.field_count()),
            fields_as_has_errors: Vec::with_capacity(self.field_count()),
            fields_as_dyn_field_ref: Vec::with_capacity(self.field_count()),
        }
    }
}

#[derive(Debug, Clone, FromField)]
#[darling(attributes(form))]
struct Field {
    ident: Option<syn::Ident>,
    ty: syn::Type,
    opt: Option<HashMap<syn::Ident, syn::Expr>>,
}

#[derive(Debug)]
struct FormDeriveBuilder {
    name: syn::Ident,
    context_struct_name: syn::Ident,
    context_struct_errors_name: syn::Ident,
    fields_as_struct_fields: Vec<TokenStream>,
    fields_as_struct_fields_new: Vec<TokenStream>,
    fields_as_context_from_request: Vec<TokenStream>,
    fields_as_from_context_vars: Vec<TokenStream>,
    fields_as_from_context: Vec<TokenStream>,
    fields_as_errors: Vec<TokenStream>,
    fields_as_errors_for: Vec<TokenStream>,
    fields_as_errors_for_mut: Vec<TokenStream>,
    fields_as_has_errors: Vec<TokenStream>,
    fields_as_dyn_field_ref: Vec<TokenStream>,
}

impl ToTokens for FormDeriveBuilder {
    fn to_tokens(&self, tokens: &mut TokenStream) {
        tokens.append_all(self.build_form_impl());
        tokens.append_all(self.build_form_context_impl());
        tokens.append_all(self.build_errors_struct());
    }
}

impl FormDeriveBuilder {
    fn push_field(&mut self, field: &Field) {
        let crate_ident = flareon_ident();
        let name = field.ident.as_ref().unwrap();
        let ty = &field.ty;
        let opt = &field.opt;

        self.fields_as_struct_fields
            .push(quote!(#name: <#ty as #crate_ident::forms::AsFormField>::Type));

        self.fields_as_struct_fields_new.push({
            let custom_options_setters: Vec<_> = if let Some(opt) = opt {
                opt.iter()
                    .map(|(key, value)| quote!(custom_options.#key = Some(#value)))
                    .collect()
            } else {
                Vec::new()
            };
            quote!(#name: {
                let options = #crate_ident::forms::FormFieldOptions {
                    id: stringify!(#name).to_owned(),
                    required: true,
                };
                type Field = <#ty as #crate_ident::forms::AsFormField>::Type;
                type CustomOptions = <Field as #crate_ident::forms::FormField>::CustomOptions;
                let mut custom_options: CustomOptions = ::core::default::Default::default();
                #( #custom_options_setters; )*
                <#ty as #crate_ident::forms::AsFormField>::new_field(options, custom_options)
            })
        });

        self.fields_as_context_from_request
            .push(quote!(stringify!(#name) => {
                #crate_ident::forms::FormField::set_value(&mut self.#name, value)
            }));

        let val_ident = format_ident!("val_{}", name);
        self.fields_as_from_context_vars.push(quote! {
            let #val_ident = <#ty as #crate_ident::forms::AsFormField>::clean_value(&context.#name).map_err(|error| {
                context.add_error(#crate_ident::forms::FormErrorTarget::Field(stringify!(#name)), error);
            })
        });
        self.fields_as_from_context
            .push(quote!(#name: #val_ident.expect("Errors should have been returned by now")));

        self.fields_as_errors
            .push(quote!(#name: Vec<#crate_ident::forms::FormFieldValidationError>));

        self.fields_as_errors_for
            .push(quote!(stringify!(#name) => self.__errors.#name.as_slice()));

        self.fields_as_errors_for_mut
            .push(quote!(stringify!(#name) => self.__errors.#name.as_mut()));

        self.fields_as_has_errors
            .push(quote!(!self.__errors.#name.is_empty()));

        self.fields_as_dyn_field_ref
            .push(quote!(&self.#name as &dyn #crate_ident::forms::DynFormField));
    }

    fn build_form_impl(&self) -> TokenStream {
        let crate_ident = flareon_ident();
        let name = &self.name;
        let context_struct_name = &self.context_struct_name;
        let fields_as_from_context_vars = &self.fields_as_from_context_vars;
        let fields_as_from_context = &self.fields_as_from_context;

        quote! {
            #[#crate_ident::__private::async_trait]
            #[automatically_derived]
            impl #crate_ident::forms::Form for #name {
                type Context = #context_struct_name;

                async fn from_request(
                    request: &mut #crate_ident::request::Request
                ) -> ::core::result::Result<#crate_ident::forms::FormResult<Self>, #crate_ident::forms::FormError> {
                    let mut context = <Self as #crate_ident::forms::Form>::build_context(request).await?;

                    use #crate_ident::forms::FormContext;
                    #( #fields_as_from_context_vars; )*

                    if context.has_errors() {
                        Ok(#crate_ident::forms::FormResult::ValidationError(context))
                    } else {
                        Ok(#crate_ident::forms::FormResult::Ok(Self {
                            #( #fields_as_from_context, )*
                        }))
                    }
                }
            }
        }
    }

    fn build_form_context_impl(&self) -> TokenStream {
        let crate_ident = flareon_ident();

        let context_struct_name = &self.context_struct_name;
        let context_struct_errors_name = &self.context_struct_errors_name;

        let fields_as_struct_fields = &self.fields_as_struct_fields;
        let fields_as_struct_fields_new = &self.fields_as_struct_fields_new;
        let fields_as_context_from_request = &self.fields_as_context_from_request;
        let fields_as_errors_for = &self.fields_as_errors_for;
        let fields_as_errors_for_mut = &self.fields_as_errors_for_mut;
        let fields_as_has_errors = &self.fields_as_has_errors;
        let fields_as_dyn_field_ref = &self.fields_as_dyn_field_ref;

        quote! {
            #[derive(::core::fmt::Debug)]
            struct #context_struct_name {
                __errors: #context_struct_errors_name,
                #( #fields_as_struct_fields, )*
            }

            #[automatically_derived]
            impl #crate_ident::forms::FormContext for #context_struct_name {
                fn new() -> Self {
                    Self {
                        __errors: ::core::default::Default::default(),
                        #( #fields_as_struct_fields_new, )*
                    }
                }

                fn fields(
                    &self,
                ) -> impl ::core::iter::DoubleEndedIterator<
                    Item = &dyn ::flareon::forms::DynFormField,
                > + ::core::iter::ExactSizeIterator + '_ {
                    [#( #fields_as_dyn_field_ref, )*].into_iter()
                }

                fn set_value(
                    &mut self,
                    field_id: &str,
                    value: ::std::borrow::Cow<str>,
                ) -> ::core::result::Result<(), #crate_ident::forms::FormFieldValidationError> {
                    match field_id {
                        #( #fields_as_context_from_request, )*
                        _ => {}
                    }
                    Ok(())
                }

                fn errors_for(
                    &self,
                    target: #crate_ident::forms::FormErrorTarget
                ) -> &[#crate_ident::forms::FormFieldValidationError] {
                    match target {
                        #crate_ident::forms::FormErrorTarget::Field(field_id) => {
                            match field_id {
                                #( #fields_as_errors_for, )*
                                _ => {
                                    panic!("Unknown field name passed to get_errors: `{}`", field_id);
                                }
                            }
                        }
                        #crate_ident::forms::FormErrorTarget::Form => {
                            self.__errors.__form.as_slice()
                        }
                    }
                }

                fn errors_for_mut(
                    &mut self,
                    target: #crate_ident::forms::FormErrorTarget
                ) -> &mut Vec<#crate_ident::forms::FormFieldValidationError> {
                    match target {
                        #crate_ident::forms::FormErrorTarget::Field(field_id) => {
                            match field_id {
                                #( #fields_as_errors_for_mut, )*
                                _ => {
                                    panic!("Unknown field name passed to get_errors_mut: `{}`", field_id);
                                }
                            }
                        }
                        #crate_ident::forms::FormErrorTarget::Form => {
                            self.__errors.__form.as_mut()
                        }
                    }
                }

                fn has_errors(&self) -> bool {
                    !self.__errors.__form.is_empty() #( || #fields_as_has_errors )*
                }
            }
        }
    }

    fn build_errors_struct(&self) -> TokenStream {
        let crate_ident = flareon_ident();
        let context_struct_errors_name = &self.context_struct_errors_name;
        let fields_as_errors = &self.fields_as_errors;

        quote! {
            #[derive(::core::fmt::Debug, ::core::default::Default)]
            struct #context_struct_errors_name {
                __form: Vec<#crate_ident::forms::FormFieldValidationError>,
                #( #fields_as_errors, )*
            }
        }
    }
}
