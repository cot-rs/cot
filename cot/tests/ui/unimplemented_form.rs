use cot::form::{Form, FormContext};

struct MyForm {}

fn main() {
    let _ = <<MyForm as Form>::Context as FormContext>::new();
}
