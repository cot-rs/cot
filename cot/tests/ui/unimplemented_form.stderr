error[E0277]: `MyForm` does not implement the `Form` trait
 --> tests/ui/unimplemented_form.rs:6:15
  |
6 |     let _ = <<MyForm as Form>::Context as FormContext>::new();
  |               ^^^^^^ `MyForm` is not a form
  |
help: the trait `cot::form::Form` is not implemented for `MyForm`
 --> tests/ui/unimplemented_form.rs:3:1
  |
3 | struct MyForm {}
  | ^^^^^^^^^^^^^
  = note: add #[derive(cot::form::Form)] to the struct to automatically derive the trait
  = help: the trait `cot::form::Form` is implemented for `DatabaseUser`
